streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -596,6 +596,109 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
       ZStream.fromQueue(queue).unTake
     }
 
+  /**
+   * Creates a stream that passes its elements through a queue that drops elements
+   * (dropping or sliding). It implements a promise that signals abrupt completion
+   * of the stream due to an error or the end of the stream. After the promise has
+   * been completed, it continues to offer the current (at the time) contents of the
+   * queue, until it has been exhausted, at which point it finishes with the signal
+   * from the promise.
+   */
+  private final def bufferSignal[A1 >: A](queue: Queue[A1]): ZStream[R, E, A1] =
+    ZStream.managed {
+      for {
+        // Signal stream error or end through the result type of the promise
+        // for better composability with other effects.
+        signal <- Promise.make[Nothing, Option[E]].toManaged_
+        _ <- self
+              .foreachManaged(a => queue.offer(a).unit)
+              .foldM(
+                e => signal.succeed(Some(e)).unit.toManaged_,
+                _ => signal.succeed(None).unit.toManaged_
+              )
+              .fork
+      } yield (queue, signal)
+    }.flatMap {
+      case (queue, signal) =>
+        ZStream.fromPull {
+          for {
+            // Promise for completing the queue take, needed so that the taken
+            // element is not lost while cleaning up other competing fibers.
+            take <- Promise.make[Nothing, A1]
+
+            // Promise that signals the availability of either a queue take
+            // or the completion of the `signal` promise.
+            latch <- Promise.make[Nothing, Unit]
+
+            // Take an element from the queue in a separate fiber.
+            takeFiber <- (queue.take.to(take) *> latch.succeed(())).fork
+
+            // Wait for either stream error or stream end.
+            signalFiber <- (signal.await *> latch.succeed(())).fork
+
+            // Park the driver fiber until either a queue take or a stream
+            // signal is available.
+            _ <- latch.await
+
+            // Check if the stream has signalled error or end. If it has,
+            // we are entering in a "queue draining" mode, in which we
+            // exhaust the values in the queue and signal completion after that.
+            a <- signal.isDone.flatMap {
+                  if (_) {
+                    // The stream has signalled error or end.
+                    take.isDone.flatMap {
+                      if (_) {
+                        // However, the queue is not empty, so we take the next element.
+                        // We don't interrupt any fibers here because they have both
+                        // completed their execution.
+                        take.await
+                      } else {
+                        // Otherwise, the queue is empty, so we interrupt the fiber
```

Comment:
Why does `take.isDone == false` mean that the queue is empty?

Reply:
Oh I understand why you went down that route. That's a race condition. There's no guarantee that the take has completed when you check it. It's safer to check the queue's size.

Reply:
Does that mean I can get rid of the `take` `Promise` as well?

Reply:
Let's see if we're finally there.